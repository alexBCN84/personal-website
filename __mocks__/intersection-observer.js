/**
 * Utility function that mocks the `IntersectionObserver` API. Necessary for components that rely
 * on it, otherwise the tests will crash. Recommended to execute inside `beforeEach`.
 * @param {object} intersectionObserverMock - Parameter that is sent to the `Object.defineProperty`
 *      overwrite method. `jest.fn()` mock functions can be passed here if the goal is to not only
 *      mock the intersection observer, but its methods.
 */
export function setupIntersectionObserverMock({
    observe = () => null,
    unobserve = () => null,
    disconnect = () => null,
  } = {}) {
    class IntersectionObserver {
      observe = observe;
      unobserve = unobserve;
      disconnect = disconnect;
    }
    Object.defineProperty(
      window,
      'IntersectionObserver',
      { writable: true, configurable: true, value: IntersectionObserver }
    );
    Object.defineProperty(
      global,
      'IntersectionObserver',
      { writable: true, configurable: true, value: IntersectionObserver }
    );
  }