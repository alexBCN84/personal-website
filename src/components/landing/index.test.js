import React from "react";
import renderer from "react-test-renderer";
import { PureHeroBanner as HeroBanner} from "./index";
import { render, cleanup, screen, waitFor } from "@testing-library/react";
import { setupIntersectionObserverMock } from "../../../__mocks__/intersection-observer"; 
import userEvent from '@testing-library/user-event';

 describe('HeroBanner', function(){

    beforeEach(setupIntersectionObserverMock);
    afterEach(cleanup); 

    let mockProps = {
      languages: { active: 'EN', inactive: 'ES' },
      setLanguage: jest.fn()
    }

    const data = {
      file: {
        childImageSharp: {
          fixed: {
            base64: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAC4jAAAuIwF4pT92AAACU0lEQVQ4y22Vz0tVURDH73u+vBFYIYhERItq3x8QEQgWZYsMikDoJ5VRq2zRxnDjRpAgqk3Zs0KJeBUtoq3VIgoKKgrLhH4aRLnQRT/o3T5D35Hxdg98mTnnzHzPnJk59yZZliU2TIJyooG+uV6vV5ET4JfwFgyDjuBXjhyRrGQ6JG3oL7N/4zcYY+0q8gq4J2Ib46y3B///yZAnZfwZ7MGhKckN1heBLvBBtn3zSHEqRzLmo4g0R3IAdIOdYKHWFoCqfPp0u7I7tOm0kUCSymit9h7psOW5wy5of2tcfIXxJydBNoTIR+RwRgVpdRvJkor1DlRsYYscdsugEnK6Rlc6jdgFbqG3eM7QK9K3i2OHLQ6rck3hRD/9rAxXgb3gJmgJ7eK3SMF3cMMmE5COFVS8VWRDmh/DruaEBdW/AyZN+Wm5Caf6NXpFeFQRHrEIQLMqvEJYKfvBTBFFQr/qYiXZx0ewX1c2wrta/2opVludcsLX4L4XRPKQHE6AjWAdOBwi7AT94JvsrLWuWbOb80XwBywJRXljzy+XI7tyzftQVbfRo/k0uJ0oAhsHtdEZ24hC+MvoRr8ufbVsHmjeoXmXn/5M+bBkPwYzoNEjDhE+gdTyOKXcL9PeOPgy92RR1uuEGg5DVoDQ5N7Em5CXdLUpf2rIc3NNnatsjzYux++iPUGbI0vKbxryel4+g952+W/hcRlY9fYxX1rQwKme2qSe5UDYK/wmbgDPRfwDPFTUVX1gZ7X3HmwLZKV5rCJtCFdtB5azpyqYRf0CjKoTKkW/gL+tuhI3CufaNQAAAABJRU5ErkJggg==",
            height: 40,
            src: "/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png",
            srcSet: "/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png 1x,/static/b0ddc18a1e874e013ee42035351d5925/183c2/logo.png 1.5x,/static/b0ddc18a1e874e013ee42035351d5925/829e2/logo.png 2x",
            width: 40
          }
        }
      },
      allDataJson: {
        edges: [
          {
            node: {
              EN: {
                description: "I've spent most of my life between Spain, France, the United Kingdom and the United States. I'm currently based in Berlin, where I work as a Frontend Engineer at Zalando SE",
                headline: "I am a former translator, a teacher at heart, and a Web Developer since 2016",
                kicker: "Welcome to my Site",
                aboutChevronText: "about us",
                cvCTA: {
                  href: "https://alexgines.netlify.app/",
                  text: "See my CV"
                }
              },
              ES: {
                aboutChevronText: "conócenos",
                cvCTA: {
                  href: "https://alexgines.netlify.app/",
                  text: "Ver CV"
                },
                description: "He pasado gran parte de mi vida entre España, Francia y Reino Unido. Actualmente vivo en Berlín, donde trabajo como Ingeniero Front-End en Zalando SE",
                headline: "Soy traductor de formación, profesor de vocación y desde 2016 desarrollador web",
                kicker: "bienvenid@ a mi web!"
              }
            }
          }
        ]
      }
    }

    it('renders desktop layout without crashing', () => {
      const {getByTestId} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByTestId('desktop')).toBeInTheDocument
    });

    // you can have as many snapshots as you want, it all depends on what props you pass into your component
    it('matches snapshot', function(){
      const tree = renderer.create(<HeroBanner  {...mockProps} data={data} />).toJSON()
      expect(tree).toMatchSnapshot()
    });
    it('displays logo appropriately', () => {
      const {getByText, getByRole} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByText('Alejandro Ginés')).toBeInTheDocument;
      expect(getByText('Web Developer')).toBeInTheDocument;
      expect(getByRole(/img/)).toBeInTheDocument;
    });
    it('shows text matching inactive language on language switch button', () => {
      const {getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByText(mockProps.languages.inactive)).toBeInTheDocument
    });
    it('fires setLanguage function when user click on language button', () => {
      expect.assertions(1);
      const {getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      const setLanguageMock = mockProps.setLanguage
      userEvent.click(getByText(mockProps.languages.inactive));
      expect(setLanguageMock).toHaveBeenCalledTimes(1);
    });
    it('shows content in English when activeLanguage is EN', () => {
      const {getByRole, getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.EN.kicker})).toBeInTheDocument;
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.EN.headline})).toBeInTheDocument;
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.EN.description})).toBeInTheDocument;
      expect(getByRole('link', {name: data.allDataJson.edges[0].node.EN.cvCTA.text})).toBeInTheDocument;
      expect(getByText(data.allDataJson.edges[0].node.EN.aboutChevronText)).toBeInTheDocument;
    });
    it('shows content in Spanish when active language is ES', () => {
      mockProps = {
        languages: { active: 'ES', inactive: 'EN' },
        setLanguage: jest.fn()
      }
      const {getByRole, getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.ES.kicker})).toBeInTheDocument;
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.ES.headline})).toBeInTheDocument;
      expect(getByRole('heading', {name: data.allDataJson.edges[0].node.ES.description})).toBeInTheDocument;
      expect(getByRole('link', {name: data.allDataJson.edges[0].node.ES.cvCTA.text})).toBeInTheDocument;
      expect(getByText(data.allDataJson.edges[0].node.ES.aboutChevronText)).toBeInTheDocument;
    });

    it('has a link to cv', () => {
      const {getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByText(data.allDataJson.edges[0].node.ES.cvCTA.text).closest('a')).toHaveAttribute('href', data.allDataJson.edges[0].node.ES.cvCTA.href);
    });

    it('mobile layout  without crashing', () => {  
      window.innerWidth = 600;
      const {getByTestId} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByTestId('mobile')).toBeInTheDocument
     });

    // test intersection Observer
    it('gives languageButton fixed position if it is in viewport', () => {
      const originalGetBoundingClientRect = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'getBoundingClientRect')
      const {getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      const switchButton = getByText(mockProps.languages.inactive)
      console.log(originalGetBoundingClientRect);
    }); 

     it('renders tablet layout without crashing', () => {
      window.innerWidth = 800;
      const {getByTestId} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
      expect(getByTestId('tablet')).toBeInTheDocument
    });
  
    // it('fires scrollTo function when about chevron is clicked', () => {
    //   window.innerWidth = 600;
    //   const {getByText} = render(<HeroBanner  {...mockProps} data={data} ></HeroBanner>)
    //   userEvent.click(getByText(data.allDataJson.edges[0].node.ES.aboutChevronText));
    //   screen.debug();
    //   console.log(document.getElementById('about').getBoundingClientRect().top)
    // });
   
 })