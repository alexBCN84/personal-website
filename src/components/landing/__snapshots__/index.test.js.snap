// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroBanner matches snapshot 1`] = `
<section
  className="sc-AxjAm cbBRmT"
>
  <div
    className="sc-fzozJi hhSIgG"
  >
    <div
      className="sc-AxirZ dCBnyz"
    >
      <div
        className="logo gatsby-image-wrapper"
        style={
          Object {
            "display": "inline-block",
            "height": 40,
            "overflow": "hidden",
            "position": "relative",
            "width": 40,
          }
        }
      >
        <img
          alt=""
          aria-hidden={true}
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAACXBIWXMAAC4jAAAuIwF4pT92AAACU0lEQVQ4y22Vz0tVURDH73u+vBFYIYhERItq3x8QEQgWZYsMikDoJ5VRq2zRxnDjRpAgqk3Zs0KJeBUtoq3VIgoKKgrLhH4aRLnQRT/o3T5D35Hxdg98mTnnzHzPnJk59yZZliU2TIJyooG+uV6vV5ET4JfwFgyDjuBXjhyRrGQ6JG3oL7N/4zcYY+0q8gq4J2Ib46y3B///yZAnZfwZ7MGhKckN1heBLvBBtn3zSHEqRzLmo4g0R3IAdIOdYKHWFoCqfPp0u7I7tOm0kUCSymit9h7psOW5wy5of2tcfIXxJydBNoTIR+RwRgVpdRvJkor1DlRsYYscdsugEnK6Rlc6jdgFbqG3eM7QK9K3i2OHLQ6rck3hRD/9rAxXgb3gJmgJ7eK3SMF3cMMmE5COFVS8VWRDmh/DruaEBdW/AyZN+Wm5Caf6NXpFeFQRHrEIQLMqvEJYKfvBTBFFQr/qYiXZx0ewX1c2wrta/2opVludcsLX4L4XRPKQHE6AjWAdOBwi7AT94JvsrLWuWbOb80XwBywJRXljzy+XI7tyzftQVbfRo/k0uJ0oAhsHtdEZ24hC+MvoRr8ufbVsHmjeoXmXn/5M+bBkPwYzoNEjDhE+gdTyOKXcL9PeOPgy92RR1uuEGg5DVoDQ5N7Em5CXdLUpf2rIc3NNnatsjzYux++iPUGbI0vKbxryel4+g952+W/hcRlY9fYxX1rQwKme2qSe5UDYK/wmbgDPRfwDPFTUVX1gZ7X3HmwLZKV5rCJtCFdtB5azpyqYRf0CjKoTKkW/gL+tuhI3CufaNQAAAABJRU5ErkJggg=="
          style={
            Object {
              "height": "100%",
              "left": 0,
              "objectFit": "cover",
              "objectPosition": "center",
              "opacity": 1,
              "position": "absolute",
              "top": 0,
              "transitionDelay": "500ms",
              "width": "100%",
            }
          }
        />
        <picture>
          <source
            srcSet="/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png 1x,/static/b0ddc18a1e874e013ee42035351d5925/183c2/logo.png 1.5x,/static/b0ddc18a1e874e013ee42035351d5925/829e2/logo.png 2x"
          />
          <img
            alt="logo"
            height={40}
            loading="lazy"
            onLoad={[Function]}
            src="/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png"
            srcSet="/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png 1x,/static/b0ddc18a1e874e013ee42035351d5925/183c2/logo.png 1.5x,/static/b0ddc18a1e874e013ee42035351d5925/829e2/logo.png 2x"
            style={
              Object {
                "height": "100%",
                "left": 0,
                "objectFit": "cover",
                "objectPosition": "center",
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "transition": "opacity 500ms",
                "width": "100%",
              }
            }
            width={40}
          />
        </picture>
        <noscript
          dangerouslySetInnerHTML={
            Object {
              "__html": "<picture><source srcset=\\"/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png 1x,/static/b0ddc18a1e874e013ee42035351d5925/183c2/logo.png 1.5x,/static/b0ddc18a1e874e013ee42035351d5925/829e2/logo.png 2x\\" /><img loading=\\"lazy\\" width=\\"40\\" height=\\"40\\" srcset=\\"/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png 1x,/static/b0ddc18a1e874e013ee42035351d5925/183c2/logo.png 1.5x,/static/b0ddc18a1e874e013ee42035351d5925/829e2/logo.png 2x\\" src=\\"/static/b0ddc18a1e874e013ee42035351d5925/ff4be/logo.png\\" alt=\\"logo\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
            }
          }
        />
      </div>
      <div
        className="sc-AxhCb ehpxdv"
      >
        <div
          className="sc-AxgMl hzwiGK"
          data-testid="name"
        >
          Alejandro Gin√©s
        </div>
        <div
          className="sc-Axmtr blfvLt"
        >
          Web Developer
        </div>
      </div>
    </div>
    <div
      className="sc-fznWqX jqXnrH"
      href="about"
    >
      <p
        className="sc-fzqARJ bqKifG"
      >
        about us
      </p>
      <div
        className="sc-fzoyTs buwGoh"
      />
    </div>
  </div>
  <div
    className="sc-fzpans fWuNxB"
  >
    <a
      className="sc-fznyAO eQYrUP"
      onClick={[Function]}
    >
      ES
    </a>
    <div
      className="sc-fznZeY deXbxg"
    >
      <h2
        className="sc-fzokOt dpyQpq"
      >
        Welcome to my Site
      </h2>
      <h1
        className="sc-fzoyAV gDGNOH"
      >
        I am a former translator, a teacher at heart, and a Web Developer since 2016
      </h1>
      <h3
        className="sc-fzpjYC enOmxh"
      >
        I've spent most of my life between Spain, France, the United Kingdom and the United States. I'm currently based in Berlin, where I work as a Frontend Engineer at Zalando SE
      </h3>
      <div
        className="wrapper"
      >
        <div
          className="link_wrapper"
        >
          <a
            className="cv_button"
            href="https://alexgines.netlify.app/"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            target="blank"
          >
            See my CV
          </a>
          <div
            className="icon"
          >
            <svg
              id="Capa_1"
              style={
                Object {
                  "height": 34,
                  "verticalAlign": "middle",
                  "width": 34,
                }
              }
              version="1.1"
              viewBox="0 0 512 512"
              x="0px"
              xmlSpace="preserve"
              y="0px"
            >
              <g>
                <g>
                  <path
                    d="M333.205,34.213L330.799,0L9.546,22.591l30.317,431.118l44.011-3.095V512h322.046v-35.276l50.265,5.413l46.27-429.698 L333.205,34.213z M83.874,420.519l-16.17,1.137L41.599,50.432l261.359-18.38l3.359,47.765H83.874V420.519z M375.9,481.979H113.895 V109.838H375.9V481.979z M429.552,449.073l-23.631-2.545V79.817h-69.508l-1.068-15.18l134.049,14.434L429.552,449.073z"
                    fill="#fff"
                  />
                </g>
              </g>
              <g>
                <g>
                  <rect
                    fill="#fff"
                    height="30.021"
                    width="195.796"
                    x="147.003"
                    y="296.746"
                  />
                </g>
              </g>
              <g>
                <g>
                  <rect
                    fill="#fff"
                    height="30.021"
                    width="195.796"
                    x="147.003"
                    y="350.143"
                  />
                </g>
              </g>
              <g>
                <g>
                  <rect
                    fill="#fff"
                    height="30.021"
                    width="195.796"
                    x="147.003"
                    y="403.54"
                  />
                </g>
              </g>
              <g>
                <g>
                  <path
                    d="M201.384,229.293c-13.238,0-22.154-9.186-22.154-23.64s8.915-23.64,22.154-23.64c8.511,0,15.13,4.053,20.668,11.078 l20.128-17.966c-9.321-12.157-23.911-18.912-42.418-18.912c-30.394,0-52.684,20.398-52.684,49.442s22.289,49.442,52.684,49.44 c18.508,0,33.097-6.754,42.418-18.912l-20.128-17.966C216.514,225.24,209.895,229.293,201.384,229.293z"
                    fill="#fff"
                  />
                </g>
              </g>
              <g>
                <g>
                  <polygon
                    fill="#fff"
                    points="321.075,158.373 297.975,213.624 275.415,158.373 241.102,158.373 281.088,252.934 312.429,252.934  352.414,158.373"
                  />
                </g>
              </g>
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
              <g />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
